src/smt.cpp

```diff
@@ -1533,155 +1551,262 @@ void Matcher::setZ3(Expr &e, optional<z3::expr> &&opt) const {
     function<bool(const Expr&)> lhsMatcher,
     function<bool(const Expr&)> rhsMatcher) const {
-  auto e = expr.getZ3Expr();
-  if (!e.is_app())
+  if (expr.hasZ3Expr()) {
+    auto e = expr.getZ3Expr();
+    if (!e.is_app())
+      return false;
+
+    Z3_app a = e;
+    Z3_func_decl decl = Z3_get_app_decl(*sctx.z3, a);
+    if (Z3_get_decl_kind(*sctx.z3, decl) != z3Kind)
+      return false;
+
+    Expr lhs = newExpr(), rhs = newExpr();
+    setZ3(lhs, z3::expr(*sctx.z3, Z3_get_app_arg(*sctx.z3, a, 0)));
+    setZ3(rhs, z3::expr(*sctx.z3, Z3_get_app_arg(*sctx.z3, a, 1)));
+    return lhsMatcher(lhs) && rhsMatcher(rhs);
+  } else {
     return false;
+  }
+}
+#endif // SOLVER_Z3
 
-  Z3_app a = e;
-  Z3_func_decl decl = Z3_get_app_decl(*sctx.z3, a);
-  if (Z3_get_decl_kind(*sctx.z3, decl) != z3Kind)
-    return false;
+#ifdef SOLVER_CVC5
+void Matcher::setCVC5(Expr &e, optional<cvc5::api::Term> &&opt) const {
+  e.setCVC5(move(opt));
+}
 
-  Expr lhs = newExpr(), rhs = newExpr();
-  setZ3(lhs, z3::expr(*sctx.z3, Z3_get_app_arg(*sctx.z3, a, 0)));
-  setZ3(rhs, z3::expr(*sctx.z3, Z3_get_app_arg(*sctx.z3, a, 1)));
-  return lhsMatcher(lhs) && rhsMatcher(rhs);
+bool Matcher::matchBinaryOp(const Expr &expr, cvc5::api::Kind opKind,
+    function<bool(const Expr&)> lhsMatcher,
+    function<bool(const Expr&)> rhsMatcher) const {
+  if (expr.hasCVC5Term()) {
+    auto term = expr.getCVC5Term();
+    if (term.getKind() != opKind || term.getNumChildren() != 2)
+      return false;
+
+    Expr lhs = newExpr(), rhs = newExpr();
+    setCVC5(lhs, move(term[0]));
+    setCVC5(rhs, move(term[1]));
+    return lhsMatcher(lhs) && rhsMatcher(rhs);
+  } else {
+    return false;
+  }
 }
-#endif // SOLVER_Z3
+
+#endif // SOLVER_CVC5
 
 
 bool ConstBool::operator()(const Expr &expr) const {
   return (val && expr.isTrue()) || (!val && expr.isFalse());
 }
 
 bool ConstSplatArray::operator()(const Expr &expr) const {
-  // FIXME: cvc5
 #ifdef SOLVER_Z3
-  auto e = expr.getZ3Expr();
-  if (!e.is_app())
-    return false;
-
-  Z3_app a = e;
-  Z3_func_decl decl = Z3_get_app_decl(*sctx.z3, a);
-  if (Z3_get_decl_kind(*sctx.z3, decl) != Z3_OP_CONST_ARRAY)
-    return false;
+  if (expr.hasZ3Expr()) {
+    auto e = expr.getZ3Expr();
+    if (!e.is_app())
+      return false;
+
+    Z3_app a = e;
+    Z3_func_decl decl = Z3_get_app_decl(*sctx.z3, a);
+    if (Z3_get_decl_kind(*sctx.z3, decl) != Z3_OP_CONST_ARRAY)
+      return false;
+
+    Expr newe = newExpr();
+    setZ3(newe, z3::expr(*sctx.z3, Z3_get_app_arg(*sctx.z3, a, 0)));
+    return subMatcher(newe);
+  }
 #endif // SOLVER_Z3
+#ifdef SOLVER_CVC5
+  if (expr.hasCVC5Term()) {
+    auto e = expr.getCVC5Term();
+
+    if (!e.isConstArray())
+      return false;
 
-  Expr newe = newExpr();
-  IF_Z3_ENABLED(setZ3(newe, z3::expr(*sctx.z3, Z3_get_app_arg(*sctx.z3, a, 0))));
-  return subMatcher(newe);
+    Expr newe = newExpr();
+    setCVC5(newe, e.getConstArrayBase());
```

Comment:
I think you should call `setCVC5` from `newe` updated at line 1616.

Reply:
Well... I think it's impossible to match Z3/CVC5 terms at the same time because the expression can be different.
So I think we should separate the implementation of Z3/CVC5 matchers on the caller's side.

Reply:
Okay.. I think the complexity comes from the fact that mlir-tv can run with two solvers (Z3, CVC5) enabled at runtime.
However, we are currently not getting any benefit from it; even if one of them finishes early, we cannot stop the other.
I will remove https://github.com/aqjune/mlir-tv/blob/17d7a7ebad2d2e000a49d5ee55dca3ee1e231dac/src/main.cpp#L53 so that only one of them is enabled at runtime. What do you think?

Reply:
Look good to me!! I'll update it.